packages/driver/src/cypress/ensure.ts

```diff
@@ -0,0 +1,314 @@
+    }
+
+    $errUtils.throwErrByPath('subject.not_element', {
+      onFail,
+      args: {
+        name,
+        subject: $utils.stringifyActual(subject),
+        previous: current.get('prev').get('name'),
+      },
+    })
+  }
+}
+
+const isWindow = (subject, name: string, cy: $Cy) => {
+  if (!$dom.isWindow(subject)) {
+    const prev = cy.state('current').get('prev')
+
+    $errUtils.throwErrByPath('subject.not_window_or_document', {
+      args: {
+        name,
+        type: 'window',
+        subject: $utils.stringifyActual(subject),
+        previous: prev.get('name'),
+      },
+    })
+  }
+}
+
+const isDocument = (subject, name: string, cy: $Cy) => {
+  if (!$dom.isDocument(subject)) {
+    const prev = cy.state('current').get('prev')
+
+    $errUtils.throwErrByPath('subject.not_window_or_document', {
+      args: {
+        name,
+        type: 'document',
+        subject: $utils.stringifyActual(subject),
+        previous: prev.get('name'),
+      },
+    })
+  }
+}
+
+const exists = (subject, cy: $Cy) => {
+  // prevent any additional logs since this is an implicit assertion
+  cy.state('onBeforeLog', () => false)
+
+  // verify the $el exists and use our default error messages
+  try {
+    cy.expect(subject).to.exist
+  } finally {
+    cy.state('onBeforeLog', null)
+  }
+}
+
+const elExists = ($el, cy: $Cy) => {
+  // ensure that we either had some assertions
+  // or that the element existed
+  if ($el && $el.length) {
+    return
+  }
+
+  // TODO: REFACTOR THIS TO CALL THE CHAI-OVERRIDES DIRECTLY
+  // OR GO THROUGH I18N
+
+  return exists($el, cy)
+}
+
+const isScrollable = ($el, name): true | void => {
+  if ($dom.isScrollable($el)) {
+    return true
+  }
+
+  const node = $dom.stringify($el)
+
+  $errUtils.throwErrByPath('dom.not_scrollable', {
+    args: { cmd: name, node },
+  })
+}
+
+/**
+  * commandCanCommunicateWithAUT will check if the command should be able to communicate with the AUT
+  * If we can not communicate, throw an error.
+  * Intended to use within retry loops.
+  * err: optional error to pass end to be appended to if the assertion happened while the aut was cross origin.
+  * @returns true or throws an error
+  */
+const commandCanCommunicateWithAUT = (cy: $Cy, err?): boolean => {
+  if (!isRunnerAbleToCommunicateWithAut()) {
+    const crossOriginCommandError = $errUtils.errByPath('miscellaneous.cross_origin_command', {
+      commandOrigin: window.location.origin,
+      autOrigin: cy.state('autLocation').origin,
+    })
+
+    if (err) {
+      err = $errUtils.appendErrMsg(err, crossOriginCommandError.message)
+
+      throw err
+    } else {
+      throw crossOriginCommandError
+    }
+  }
+
+  return true
+}
+
+export default {
+  isElement,
+  isAttached,
+  isWindow,
+  isDocument,
+  isNotDisabled,
+  isNotReadonly,
+  isVisible,
+  isStrictlyVisible,
+  isNotHiddenByAncestors,
+  exists,
+  elExists,
+  isScrollable,
+  commandCanCommunicateWithAUT,
+
+  // internal functions
+  isType,
```

Comment:
`isType` is listed as internal, but has public typescript types. do types need removed or can we updat?

Reply:
Redid / reordered this list, both to specify which ones are internal and to match the order of type definitions. :+1: